### PR TITLE
Support parsing for multiline string

### LINKED ISSUE #46: Multiline literal strings
Multi-line literal strings are surrounded by three single quotes on each side and allow newlines. Like literal strings, there is no escaping whatsoever. A newline immediately following the opening delimiter will be trimmed. All other content between the delimiters is interpreted as-is without modification.
```
regex2 = '''I [dw]on't need \d{2} apples'''
lines  = '''
The first newline is
trimmed in raw strings.
   All other whitespace
   is preserved.
'''
```

### PR DESCRIPTION
**Pull request by BOOMeranGG**
Fixes #46

### CODE REVIEW

**Comment by akuleshov7** [comment]
@NightEule5 you will get conflicts with this one PR, I guess… 🥲

**Comment by NightEule5** [comment]
> @NightEule5 you will get conflicts with this one PR, I guess… smiling_face_with_tear

Ok with me

File: ktoml-core/src/commonTest/kotlin/com/akuleshov7/ktoml/decoders/multiline/BasicMultilineStringDecoderTest.kt [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+            second line \
+            
+            third line""${'"'}
+        """.trimIndent()
+        assertEquals(SimpleString("first line second line third line"), Toml.decodeFromString(test))
+
+        test = "a = \"\"\"\n" +
+            "The quick brown \\\n\n\n" +
+            "  fox jumps over \\\n" +
+            "    the lazy dog.\"\"\""
```
Comment (NightEule5) [inline ktoml-core/src/commonTest/kotlin/com/akuleshov7/ktoml/decoders/multiline/BasicMultilineStringDecoderTest.kt:93]: Why not use multiline strings here? If it's the quotes, when I have to write triple quotes I usually use templates similar to what you did above this. The strings here are bit hard to read

Something like this should work:

```kotlin
val tripleQuotes = "\"\"\""

test = """
a = $tripleQuotes
The quick brown \



  fox jumps over \
    the lazy dog.$tripleQuotes
""".trimIndex()
```

**Comment by NightEule5** [comment]
Nothing else to add, looks good to me. We'll see what akuleshov7 thinks, parsing is more his thing

File: ktoml-core/src/commonTest/kotlin/com/akuleshov7/ktoml/decoders/multiline/BasicMultilineStringDecoderTest.kt (hunk 1 above)
Comment (BOOMeranGG) [reply ktoml-core/src/commonTest/kotlin/com/akuleshov7/ktoml/decoders/multiline/BasicMultilineStringDecoderTest.kt]: Added

File: ktoml-core/src/commonMain/kotlin/com/akuleshov7/ktoml/tree/nodes/pairs/TomlKeyValue.kt [hunk 2]
```diff
@@ -138,13 +138,7 @@ public fun String.parseValue(lineNo: Int, config: TomlInputConfig): TomlValue =
...
     // ===== strings
     else -> when (this.first()) {
-        '\'' -> if (this.startsWith("'''")) {
-            // actually the support for multiline strings will be here, but now just a simple mock
-            // this mock will raise an exception here
-            TomlBasicString(this, lineNo)
-        } else {
-            TomlLiteralString(this, lineNo, config)
-        }
+        '\'' -> TomlLiteralString(this, lineNo, config)
```
Comment (akuleshov7) [inline ktoml-core/src/commonMain/kotlin/com/akuleshov7/ktoml/tree/nodes/pairs/TomlKeyValue.kt:141]: So we received a multiline string and are parsing it with `TomlLiteralString`? 
It's ok, if it's not a mistake, but while reading the code - I would expect here different types of strings that start from `'`: Literal, Multiline, etc. 

As this code was intended to be an entry point for determination of different types.

File: ktoml-core/src/commonMain/kotlin/com/akuleshov7/ktoml/tree/nodes/pairs/TomlKeyValue.kt [hunk 3]
```diff
@@ -138,13 +138,7 @@ public fun String.parseValue(lineNo: Int, config: TomlInputConfig): TomlValue =
...
         // ===== basic strings
         '\"' -> TomlBasicString(this, lineNo)
```
Comment (akuleshov7) [inline ktoml-core/src/commonMain/kotlin/com/akuleshov7/ktoml/tree/nodes/pairs/TomlKeyValue.kt:143]: So we received a multiline string and are parsing it with `TomlBasicString`? 
It's ok, if it's not a mistake, but while reading the code - I would expect here different types of strings that start from `"`: Basic, Multiline, etc. 

As this code was intended to be an entry point for determination of different types. So I got a little bit confused.

File: ktoml-core/src/commonMain/kotlin/com/akuleshov7/ktoml/tree/nodes/pairs/values/TomlBasicString.kt [hunk 4]
```diff
@@ -43,19 +46,25 @@ public class TomlBasicString internal constructor(
     public companion object {
         private fun String.verifyAndTrimQuotes(lineNo: Int): Any =
                 when {
+                    // ====== multiline string (""") =======
```
Comment (akuleshov7) [inline ktoml-core/src/commonMain/kotlin/com/akuleshov7/ktoml/tree/nodes/pairs/values/TomlBasicString.kt:49]: Yeah, I see: you decided not to make differences on the higher level. 
That's actually okay, may I should have done the same with LiteralStrings

**Comment by akuleshov7** [comment]
@BOOMeranGG I think we have missed multiline literal strings also:

image

Can be done separately, np

**Review by akuleshov7 — APPROVED** [review]
Looked more closely, actually awesome work, lgtm!
Thank you.

PS great work with tests also

File: ktoml-core/src/commonMain/kotlin/com/akuleshov7/ktoml/parsers/TomlParser.kt [hunk 5]
```diff
@@ -135,52 +128,91 @@ public value class TomlParser(private val config: TomlInputConfig) {
...
         val firstEqualsSign = line.indexOfFirst { it == '=' }
         if (firstEqualsSign == -1) {
-            return false
+            return MultilineType.NOT_A_MULTILINE
         }
         val value = line.substring(firstEqualsSign + 1).trim()
 
-        return value.startsWith("[") &&
-                value.endsWith("]").not()
+        if (value.startsWith("[") && !value.endsWith("]")) {
```
Comment (akuleshov7) [inline ktoml-core/src/commonMain/kotlin/com/akuleshov7/ktoml/parsers/TomlParser.kt:188]: the only stylistic comment I have here - is usage of `return when { 

}
`

It usually looks much better with a huge number of `if statements`

**Pull request merged**